lib/resty/openssl/x509/store_ctx.lua

```diff
@@ -0,0 +1,234 @@
+  end
+
+  local store_ctx
+  if store then
+    store_ctx = store.ctx
+  end
+
+  local x509_ctx
+  if x509 then
+    x509_ctx = x509.ctx
+  end
+
+  if C.X509_STORE_CTX_init(ctx, store_ctx, x509_ctx, chain_dup_ctx) ~= 1 then
+    return nil, format_error("x509.store_ctx:init: X509_STORE_CTX_init")
+  end
+
+  return setmetatable({
+    ctx = ctx,
+    verified = false,
+  }, mt), nil
+end
+
+function _M.istype(l)
+  return l and l.ctx and ffi.istype(store_ctx_ptr_ct, l.ctx)
+end
+
+function _M:set_default(verify_method)
+  if verify_method and C.X509_STORE_CTX_set_default(self.ctx, verify_method) ~= 1 then
+    return nil, "x509.store_ctx:set_default: invalid verify_method \"" .. verify_method .. "\""
+  end
+
+  return true
+end
+
+function _M:set_purpose(purpose)
+  if type(purpose) ~= "string" then
+    return nil, "x509.store:set_purpose: expect a string at #1"
+  end
+
+  local pchar = ffi.new("char[?]", #purpose, purpose)
+  local idx = C.X509_PURPOSE_get_by_sname(pchar)
+  idx = tonumber(idx)
+
+  if idx == -1 then
+    return nil, "invalid purpose \"" .. purpose .. "\""
+  end
+
+  local purp = C.X509_PURPOSE_get0(idx)
+  local i = C.X509_PURPOSE_get_id(purp)
+
+  if C.X509_STORE_CTX_set_purpose(self.ctx, i) ~= 1 then
+    return false, format_error("x509.store_ctx:set_purpose: X509_STORE_set_purpose")
+  end
+
+  return true
+end
+
+function _M:set_flags(...)
+  local flags = 0
+  for _, f in ipairs({...}) do
+    flags = bor(flags, f)
+  end
+
+  C.X509_STORE_CTX_set_flags(self.ctx, flags)
+
+  return true
+end
+
+function _M:add_crl(crl)
+  if not crl_lib.istype(crl) then
+    return nil, "x509.store_ctx:add_crl: expect a x509.crl instance at #3"
+  end
+
+  if not self.crl_stack then
+    local raw, err = crl_stack_new()
+    if raw == nil then
+      return nil, err
+    end
+
+    self.crl_stack = setmetatable({
+      ctx = raw,
+    }, crl_stack_mt)
+  end
+
+  local dup = C.X509_CRL_dup(crl.ctx)
+  if dup == nil then
+    return nil, format_error("x509.store_ctx:add_crl: X509_CRL_dup")
+  end
+
+  local ok, err = crl_stack_add(self.crl_stack.ctx, dup)
+  if not ok then
+    C.X509_CRL_free(dup)
+    return nil, err
+  end
+
+  return true
+end
+
+function _M:verify(return_chain)
+  if self.crl_stack then
+    C.X509_STORE_CTX_set0_crls(self.ctx, self.crl_stack.ctx)
+
+    -- auto-set the flag
+    self:set_flags(flag_crl_check)
+  end
+
+  local code = C.X509_verify_cert(self.ctx)
+  if code == 1 then -- verified
+    self.verified = true
+    if not return_chain then
+      return true, nil
+    end
+
+    local chain_ctx = x509_vfy_macro.X509_STORE_CTX_get0_chain(self.ctx)
+    return chain_lib.dup(chain_ctx)
+
+  elseif code == 0 then -- unverified
+    local vfy_code = C.X509_STORE_CTX_get_error(self.ctx)
+    return nil, ffi_str(C.X509_verify_cert_error_string(vfy_code))
+  end
+
+  -- error
+  return nil, format_error("x509.store_ctx:verify: X509_verify_cert", code)
+end
+
+function _M:check_revocation(verified_chain)
```

Comment:
this function seems not used by the caller `store:verify`, is it needed?

Reply:
`store:verify` always build and check the chain and also does revocation checking if the flag `X509_V_FLAG_CRL_CHECK` is set. But it doesn't support only doing revocation checking. `check_revocation` as a complement, only does the revocation checking with the already verified chain.